### DISCUSSION TITLE
Help with importing/installing Windows

### DISCUSSION THREAD

**Discussion by AnestisK** [cobbler/cobbler#3490] [General]
Hi,

We have a cobbler setup, and it works (so far) with Linux (specifically Fedora) systems.

We are trying to get Windows installation via Cobbler working and we are experiencing difficulties, and the wingen.rst documentation is very confusing and doesn't provide a clear easy step-by step process, so I'm looking for some help and possibly a nice easy step-by-step guide to be able to, initially, install Windows 10 via Cobbler.

i.e.

1.  Do this
2. Do this step next
3. After step 3 do step 4

etc., etc.

Do note that our DHCP server is Windows based.

We did try using the import cobbler command, ` import --name=Win10Custom --source=/mnt `, and the import worked and the profile shows up in the PXE Boot menu, however, selecting the profile does nothing, and pressing tab shows the command as `/var/lib/tftpboot/images/Win10Custom/pxeboot.n12` as the command.

Preferably, we would like to use iPXE as well.

Can anyone assist?

Cobbler version 3.3.3 on Fedora 38 with SELinux enabled.

**Comment by SchoolGuy** [comment]
Have a look at this comment: https://github.com/cobbler/cobbler/issues/3473#issuecomment-1731821253

Or just read the docs: https://cobbler.readthedocs.io/en/v3.3.3/user-guide/wingen.html

**Comment by AnestisK** [comment]
> Have a look at this comment: [#3473 (comment)](https://github.com/cobbler/cobbler/issues/3473#issuecomment-1731821253)
> 
> Or just read the docs: https://cobbler.readthedocs.io/en/v3.3.3/user-guide/wingen.html

The docs do not provide a good clear step-by-step process.  They jump all over the place.

As mentioned, I followed the docs and did the basic import mentioned at the very beginning (which, in my opinion should just work), and I just end up back to the PXE Boot menu when selecting the profile.

An import should just work without having to perform any additional/ complicated configuration steps.

**Reply by SchoolGuy** [reply]
@tpw56j Sorry to bother you repeatedly but it seems people are sadly still not happy with your instructions. Could you try and help out again, please?

**Reply by tpw56j** [reply]
Unfortunately, when using cobbler 3.3.3 from the Fedora 38, there is probably no simpler way than those described in https://github.com/cobbler/cobbler/issues/3473#issuecomment-1731821253. The fixes needed for UEFI were only made in PR #3477 and are not yet included in the Fedora distro.

In the same PR, the documentation was rewritten (https://github.com/cobbler/cobbler/blob/release33/docs/user-guide/wingen.rst), and I hope that it has become more useful.

**Comment by AnestisK** [comment]
Well, from the original link provided I was attempting thsi step:
`cobbler distro add --name=Win10Custom2 --kernel=/var/lib/tftpboot/winos/Win10Custom-x86_64/pxeboot.n12 --initrd=/var/lib/tftpboot/winos/Win10Custom-x86_64/boot/boot.sdi --boot-loader=pxe --arch=x86_64 --breed=windows --os-version=10`

and got this response:

`exception on server: initrd not found`

The path and files definitely exist in the locations listed.

**Comment by tpw56j** [comment]
In version 3.3.3, the `/var/lib/tftpboot/winos` directory is no longer used for the location of windows distributions. Instead, windows distros should be placed in the `/var/www/cobbler/distro_mirror` directory. The `cobbler import` command copies windows files exactly there. Additionally, `cobbler import` does the following:
1. extracts the WinPE image from the `sources/boot.wim` into the `boot/winpe.wim`
2. extracts the file `/Windows/Boot/PXE/pxeboot.n12` from `boot/winpe.wim` to `boot/pxeboot.n12`
3. extracts the file `/Windows/Boot/PXE/bootmgr.exe` from `boot/winpe.wim` to `boot/bootmgr.exe`
4. replaces all substrings `X:\\$windows.~bt` with `X:` in `winpe.wim:/Windows/System32/config/SOFTWARE` registry keys and values for PXE boot

Therefore, instead of the `cobbler distro add` command, I suggest using the following:
1. `mount -o loop,ro Win10_22H2_English_x64v1.iso /mnt`
2. `cobbler import --name=Win10Custom --path=/mnt`
3. `cobbler sync`

A working version of cobbler can be found here: https://github.com/cobbler/cobbler/tree/release33
current windows documentation: https://github.com/cobbler/cobbler/blob/release33/docs/user-guide/wingen.rst

**Reply by AnestisK** [reply]
I have done everything you have suggested above and get the issue I mentioned where selecting the Windows profile in the PXE Boot menu results in nothing happening and the system returning back to the PXE Boot menu.
Next time please actually read the actual issue instead of providing a boilerplate response.

**Reply by tpw56j** [reply]
Have you followed all the steps from https://github.com/cobbler/cobbler/issues/3473#issuecomment-1731821253?

What I suggested above is only part of the actions that need to be performed instead of the `cobbler distro add` that you used, which is unacceptable in the case of windows.

I also note that neither the documentation https://github.com/cobbler/cobbler/tree/release33 nor the more detailed instructions in https://github.com/cobbler/cobbler/issues/3473#issuecomment-1731821253 indicate that you can use the `cobbler distro add` to configure the installation of Windows 10 .

**Reply by AnestisK** [reply]
And if you had read everything fully, you would have seen that I had tried the initial configuration in the original Cobbler 3.3.3 documentation which uses both cobbler import and cobbler distro add.
Again, you would know this if you actually read my issue, the initial cobbler documentation which I referred to, as well as the first set of instructions linked to.
The new wingen.rst that someone else provided actually has better documentation.  Still not perfect, but better.  And so far seems to work better.
However, it assumes that one is using Linux DHCP and provides no information on how to set up Windows DHCP for various booting options (UEFI, BIOS, iPXE, PXE, and various combinations of as mentioned in the new wingen.rst file).

**Comment by AnestisK** [comment]
So, I've followed the isntructions at https://github.com/cobbler/cobbler/blob/release33/docs/user-guide/wingen.rst (PXE Boot with BIOS), and we get a boot but the network share fails to load.
Yes, I have configured SAMBA and tftp.rules and changed the tftp service file to include the rules as per the instructions.
But I now get the following issues:
During the boot process I get the following error for the network share load process (where [ipaddress is our test server's ip address]):
```
net use z: \\[ipaddres]\DISTRO\win10-x86_64
System error 67 has occured
The network name cannot be found
```
The following show for both tftp.service and smb.service:
```
Oct 30 08:03:49 cobbler-test in.tftpd[33054]: Client ::ffff:159.60.34.145 finished libutil.c32
Oct 30 08:03:49 cobbler-test in.tftpd[33055]: Client ::ffff:159.60.34.145 finished pxelinux.cfg/default
Oct 30 08:03:52 cobbler-test in.tftpd[33056]: Client ::ffff:159.60.34.145 finished /images/win10-x86_64/win10a.0
Oct 30 08:03:53 cobbler-test in.tftpd[33057]: Client ::ffff:159.60.34.145 finished boot1ea.exe
Oct 30 08:03:53 cobbler-test in.tftpd[33059]: tftp: client does not accept options
Oct 30 08:03:53 cobbler-test in.tftpd[33060]: Client ::ffff:159.60.34.145 finished \Boot\1Ea
Oct 30 08:03:53 cobbler-test in.tftpd[33063]: tftp: client does not accept options
Oct 30 08:03:53 cobbler-test in.tftpd[33064]: tftp: client does not accept options
Oct 30 08:03:53 cobbler-test in.tftpd[33067]: Client ::ffff:159.60.34.145 finished \images\win10-x86_64\boot.sdi
Oct 30 08:04:16 cobbler-test in.tftpd[33068]: Client ::ffff:159.60.34.145 finished \images\win10-x86_64\winp5.wim
```
```
Oct 30 08:03:49 cobbler-test in.tftpd[33054]: Client ::ffff:159.60.34.145 finished libutil.c32
Oct 30 08:03:49 cobbler-test in.tftpd[33055]: Client ::ffff:159.60.34.145 finished pxelinux.cfg/default
Oct 30 08:03:52 cobbler-test in.tftpd[33056]: Client ::ffff:159.60.34.145 finished /images/win10-x86_64/win10a.0
Oct 30 08:03:53 cobbler-test in.tftpd[33057]: Client ::ffff:159.60.34.145 finished boot1ea.exe
Oct 30 08:03:53 cobbler-test in.tftpd[33059]: tftp: client does not accept options
Oct 30 08:03:53 cobbler-test in.tftpd[33060]: Client ::ffff:159.60.34.145 finished \Boot\1Ea
Oct 30 08:03:53 cobbler-test in.tftpd[33063]: tftp: client does not accept options
Oct 30 08:03:53 cobbler-test in.tftpd[33064]: tftp: client does not accept options
Oct 30 08:03:53 cobbler-test in.tftpd[33067]: Client ::ffff:159.60.34.145 finished \images\win10-x86_64\boot.sdi
Oct 30 08:04:16 cobbler-test in.tftpd[33068]: Client ::ffff:159.60.34.145 finished \images\win10-x86_64\winp5.wim
```

**Reply by AnestisK** [reply]
And we managed to fix that issue but we are now getting this:
```
[2023/10/30 13:30:31.196881, 10, pid=4155, effective(65534, 65534), real(65534, 0), class=smb2] ../../source3/smbd/smb2_create.c:748(smbd_smb2_create_send)
  smbd_smb2_create_send: name [win10-x86_64\sources\setup.exe]
[2023/10/30 13:30:31.196943, 10, pid=4155, effective(65534, 65534), real(65534, 0), class=smb2] ../../source3/smbd/smb2_create.c:864(smbd_smb2_create_send)
  smbd_smb2_create_send: open execution phase
[2023/10/30 13:30:31.197331,  3, pid=4155, effective(65534, 65534), real(65534, 0), class=smb2] ../../source3/smbd/smb2_server.c:3962(smbd_smb2_request_error_ex)
  smbd_smb2_request_error_ex: smbd_smb2_request_error_ex: idx[1] status[NT_STATUS_ACCESS_DENIED] || at ../../source3/smbd/smb2_create.c:340
[2023/10/30 13:30:31.197408, 10, pid=4155, effective(65534, 65534), real(65534, 0), class=smb2] ../../source3/smbd/smb2_server.c:3847(smbd_smb2_request_done_ex)
  smbd_smb2_request_done_ex: mid [19] idx[1] status[NT_STATUS_ACCESS_DENIED] body[8] dyn[yes:1] at ../../source3/smbd/smb2_server.c:4013
```
Any help with setting up Samba to allow guest authentication/no password would help, and also what's required in the Windows Image or Startnet template appreciated.

**Reply by AnestisK** [reply]
I found the answer.  This needs to be added to the DISTRO smb share in smb.conf:
`	acl allow execute always = yes`

**Comment by AnestisK** [comment]
The following line needs to be added to the '[DISTRO]' section in `/etc/samba/smb.conf` for the Cobbler share: 
`acl allow execute always = yes`

**Reply by badacook** [reply]
Hi AnestisK 
      As you say you had finished os deployment platform via Cobbler version 3.3.3 on Fedora 38, which released version had been automatic installation successfully in production environment, and looking forward to your sharing more experience , thanks.
